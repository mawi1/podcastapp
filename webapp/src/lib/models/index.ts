import { type Result, err, ok } from "neverthrow";

export type PodcastBaseData = {
  id: number;
  title: string;
  latestEpisode: Date | null;
};

export type PodcastRef = {
  id: number;
  title: string;
};

export type EpisodeBaseData = {
  id: number;
  title: string | null;
  pubDate: Date;
  durationSeconds: number | null;
};

export type ResumeData = {
  currentTime: number,
  playbackRate: number,
}

export type PaginatedList<T> = {
  items: T[];
  totalItems: number;
};

export class PageNo {
  private _p: number;

  private constructor(p: number) {
    this._p = p;
  }

  public static fromUrl(url: URL): Result<PageNo, "invalid_page_no"> {
    const param = url.searchParams.get("page");
    if (param === null) {
      return ok(new PageNo(1));
    }
    const n = parseInt(param);
    if (isNaN(n) || n < 1 || n > Number.MAX_SAFE_INTEGER) {
      return err("invalid_page_no");
    } else {
      return ok(new PageNo(n));
    }
  }

  get pageNo(): number {
    return this._p;
  }
}

export class Id {
  private _id: number;

  private constructor(id: number) {
    this._id = id;
  }

  public static fromStr(idStr: string): Result<Id, "invalid_id"> {
    const n = parseInt(idStr);
    if (isNaN(n) || n < 0 || n > Number.MAX_SAFE_INTEGER) {
      return err("invalid_id");
    } else {
      return ok(new Id(n));
    }
  }

  get id(): number {
    return this._id;
  }
}

export enum InitialSymbol {
  A = "A",
  B = "B",
  C = "C",
  D = "D",
  E = "E",
  F = "F",
  G = "G",
  H = "H",
  I = "I",
  J = "J",
  K = "K",
  L = "L",
  M = "M",
  N = "N",
  O = "O",
  P = "P",
  Q = "Q",
  R = "R",
  S = "S",
  T = "T",
  U = "U",
  V = "V",
  W = "W",
  X = "X",
  Y = "Y",
  Z = "Z",
  Numeric = "0-9",
  Other = "Other",
}

export function parseInitialSymbol(val: string): undefined | InitialSymbol {
  const match = Object.entries(InitialSymbol).find(([, value]) => value === val);
  if (match) {
    return match[1];
  } else {
    return undefined;
  }
}
