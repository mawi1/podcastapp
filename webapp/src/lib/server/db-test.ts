import { type NodePgDatabase, drizzle } from "drizzle-orm/node-postgres";
import { migrate } from "drizzle-orm/postgres-js/migrator";
import path, { dirname } from "path";
import { Client, Pool } from "pg";
import sha1 from "simple-sha1";
import { fileURLToPath } from "url";
import { expect, inject, test, vi } from "vitest";

import * as schema from "./schema";
import { episode, podcast } from "./schema";

export type DB = NodePgDatabase<typeof import("./schema")> & { $client: Pool };

export async function insertTestUser(db: DB): Promise<void> {
  const passwordHash =
    "$argon2id$v=19$m=19456,t=2,p=1$C4ma4F2Ing8dghbBKekbmw$L6K7xP4lGY2jI79nheDwCK1QXYiKShcJtOJ4t4xdRBg"; // password111
  await db.insert(schema.user).values({ username: "testuser", passwordHash });
}

export async function insertTestPodcasts(db: DB) {
  const insert = async (title: string): Promise<number> => {
    const link = `http://${title.replaceAll(" ", "").toLocaleLowerCase()}.com`;
    const feedUrl = link + "/feed.xml";
    const description = `Podcast ${title} description`;
    const podcastId = (
      await db
        .insert(podcast)
        .values({ title: title, link, description, feedUrl })
        .returning({ id: podcast.id })
    )[0].id;
    return podcastId;
  };

  const podcasts = [
    {
      title: "The Startup Squad",
      episodes: 34,
      // id 1
    },
    {
      title: "Love in the Algorithm",
      episodes: 101,
      // id 2
    },
    {
      title: "Murder at the Microphone",
      episodes: 45,
      // id 3
    },
    {
      title: "The Paranormal Pundits",
      episodes: 29,
      // id 4
    },
    {
      title: "Tech Tales from the Future",
      episodes: 255,
      // id 5
    },
    {
      title: "Fitness Frontier",
      episodes: 12,
      // id 6
    },
    {
      title: "Culinary Crime Scene",
      episodes: 22,
      // id 7
    },
    {
      title: "The Historical Haunts",
      episodes: 612,
      // id 8
    },
    {
      title: "8-Bit Adventures",
      episodes: 93,
      // id 9
    },
    {
      title: "12 Steps to Success",
      episodes: 278,
      // id 10
    },
    {
      title: "@Tech Talk ",
      episodes: 145,
      // id 11
    },
    {
      title: "+Positive Vibes",
      episodes: 102,
      // id 12
    },
  ];
  const startDate = new Date(Date.UTC(2024, 1, 1, 0, 0, 0, 0)).getTime();
  for (const p of podcasts) {
    const podcastId = await insert(p.title);
    for (let i = 1; i <= p.episodes; i++) {
      const title = `Podcast (${podcastId}) Episode #${i}`;
      const content = "<p>Content ${title}</p>";
      const audioUrl = `http://podcastapp.mawi.cc/podcast/${podcastId}/${i}/audio.mp3`;
      const guid = `${podcastId}_${i}`;
      const durationSeconds = 234 + 3 * i;
      const pubDate = new Date(startDate + podcastId * 8 + i * 3600000 * 17);
      await db
        .insert(episode)
        .values({ podcastId, title, content, audioUrl, guid, durationSeconds, pubDate });
    }
  }
}

interface Fixture {
  db: DB;
}

type PostgresConfig = {
  user: string;
  password: string;
  host: string;
  port: number;
};

async function run_sql(config: PostgresConfig, sql: string) {
  const client = new Client(config);
  await client.connect();
  client.on("error", (err) => {
    console.error("something bad has happened!", err.stack);
  });
  await client.query(sql);
  await client.end();
}

declare module "vitest" {
  export interface ProvidedContext {
    postgresConfig: {
      user: string;
      password: string;
      host: string;
      port: number;
    };
  }
}

export const dbTest = test.extend<Fixture>({
  // eslint-disable-next-line no-empty-pattern
  db: async ({}, use) => {
    vi.resetModules();

    const config = inject("postgresConfig");

    if (expect.getState().currentTestName === undefined) {
      throw "no test name provided";
    }
    const dbName = "D" + sha1.sync(expect.getState().currentTestName!);
    await run_sql(config, `CREATE DATABASE "${dbName}";`);

    const migrationsFolder = path.resolve(
      dirname(fileURLToPath(import.meta.url)),
      "..",
      "..",
      "..",
      "drizzle"
    );
    const pool = new Pool({
      user: config.user,
      password: config.password,
      host: config.host,
      port: config.port,
      database: dbName,
    });
    const db = drizzle(pool, { schema });
    await migrate(db, { migrationsFolder });

    vi.doMock("./db.ts", () => {
      return {
        default: db,
      };
    });
    await use(db);
    await pool.end();
    await run_sql(config, `DROP DATABASE "${dbName}";`);
  },
});
