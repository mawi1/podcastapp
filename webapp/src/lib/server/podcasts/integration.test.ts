import { describe, expect } from "vitest";

import { Id, InitialSymbol } from "$lib/models";

import { dbTest, insertTestPodcasts, insertTestUser } from "../db-test";

describe("podcasts", () => {
  dbTest("get by id", async ({ db }) => {
    const { getPodcastById } = await import(".");

    await insertTestUser(db);
    await insertTestPodcasts(db);

    const p = await getPodcastById(1, Id.fromStr("4")._unsafeUnwrap());
    expect(p).toMatchSnapshot();
  });

  dbTest("get by id subscribed", async ({ db }) => {
    const { getPodcastById } = await import(".");
    const { subscribe } = await import("../subscription");

    await insertTestUser(db);
    await insertTestPodcasts(db);
    await subscribe(1, 3);

    const p = await getPodcastById(1, Id.fromStr("3")._unsafeUnwrap());
    expect(p).toMatchSnapshot();
  });

  dbTest("get by initial symbol alphabetic", async ({ db }) => {
    const { getPodcastsByInitialSymbol } = await import(".");
    const { subscribe } = await import("../subscription");

    await insertTestUser(db);
    await insertTestPodcasts(db);
    await subscribe(1, 8);

    const p = await getPodcastsByInitialSymbol(1, InitialSymbol.T);
    expect(p).toMatchSnapshot();
  });

  dbTest("get by initial symbol numeric", async ({ db }) => {
    const { getPodcastsByInitialSymbol } = await import(".");
    const { subscribe } = await import("../subscription");

    await insertTestUser(db);
    await insertTestPodcasts(db);
    await subscribe(1, 10);

    const p = await getPodcastsByInitialSymbol(1, InitialSymbol.Numeric);
    expect(p).toMatchSnapshot();
  });

  dbTest("get by initial symbol other", async ({ db }) => {
    const { getPodcastsByInitialSymbol } = await import(".");
    const { subscribe } = await import("../subscription");

    await insertTestUser(db);
    await insertTestPodcasts(db);
    await subscribe(1, 12);

    const p = await getPodcastsByInitialSymbol(1, InitialSymbol.Other);
    expect(p).toMatchSnapshot();
  });

  dbTest("search empty query string", async ({ db }) => {
    const { searchPodcasts } = await import(".");

    await insertTestUser(db);
    await insertTestPodcasts(db);

    const p = await searchPodcasts(1, "");
    expect(p.length).toBe(0);
  });

  dbTest("search", async ({ db }) => {
    const { searchPodcasts } = await import(".");
    const { subscribe } = await import("../subscription");

    await insertTestUser(db);
    await insertTestPodcasts(db);
    await subscribe(1, 5);

    const p = await searchPodcasts(1, "tech");
    expect(p).toMatchSnapshot();
  });
});
