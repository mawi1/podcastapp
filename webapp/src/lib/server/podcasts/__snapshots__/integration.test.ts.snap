// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`podcasts > get by id 1`] = `
{
  "description": "Podcast The Paranormal Pundits description",
  "id": 4,
  "isSubscribed": false,
  "link": "http://theparanormalpundits.com",
  "title": "The Paranormal Pundits",
}
`;

exports[`podcasts > get by id subscribed 1`] = `
{
  "description": "Podcast Murder at the Microphone description",
  "id": 3,
  "isSubscribed": true,
  "link": "http://murderatthemicrophone.com",
  "title": "Murder at the Microphone",
}
`;

exports[`podcasts > get by initial symbol alphabetic 1`] = `
[
  {
    "id": 5,
    "isSubscribed": false,
    "latestEpisode": 2024-07-30T15:00:00.040Z,
    "title": "Tech Tales from the Future",
  },
  {
    "id": 8,
    "isSubscribed": true,
    "latestEpisode": 2025-04-09T12:00:00.064Z,
    "title": "The Historical Haunts",
  },
  {
    "id": 4,
    "isSubscribed": false,
    "latestEpisode": 2024-02-21T13:00:00.032Z,
    "title": "The Paranormal Pundits",
  },
  {
    "id": 1,
    "isSubscribed": false,
    "latestEpisode": 2024-02-25T02:00:00.008Z,
    "title": "The Startup Squad",
  },
]
`;

exports[`podcasts > get by initial symbol numeric 1`] = `
[
  {
    "id": 10,
    "isSubscribed": true,
    "latestEpisode": 2024-08-15T22:00:00.080Z,
    "title": "12 Steps to Success",
  },
  {
    "id": 9,
    "isSubscribed": false,
    "latestEpisode": 2024-04-06T21:00:00.072Z,
    "title": "8-Bit Adventures",
  },
]
`;

exports[`podcasts > get by initial symbol other 1`] = `
[
  {
    "id": 12,
    "isSubscribed": true,
    "latestEpisode": 2024-04-13T06:00:00.096Z,
    "title": "+Positive Vibes",
  },
  {
    "id": 11,
    "isSubscribed": false,
    "latestEpisode": 2024-05-13T17:00:00.088Z,
    "title": "@Tech Talk ",
  },
]
`;

exports[`podcasts > search 1`] = `
[
  {
    "id": 5,
    "isSubscribed": true,
    "latestEpisode": 2024-07-30T15:00:00.040Z,
    "title": "Tech Tales from the Future",
  },
  {
    "id": 11,
    "isSubscribed": false,
    "latestEpisode": 2024-05-13T17:00:00.088Z,
    "title": "@Tech Talk ",
  },
]
`;
